lib/ModelMixins/CatalogFunctionJobMixin.ts

```diff
@@ -0,0 +1,316 @@
+     */
+    protected abstract async _invoke(): Promise<boolean>;
+
+    public async invoke() {
+      this.setTrait(CommonStrata.user, "jobStatus", "running");
+      try {
+        const finished = await this._invoke();
+        if (finished) {
+          this.setTrait(CommonStrata.user, "jobStatus", "finished");
+          this.onJobFinish(true);
+        } else {
+          this.setTrait(CommonStrata.user, "refreshEnabled", true);
+        }
+      } catch (error) {
+        this.setTrait(CommonStrata.user, "jobStatus", "error");
+        this.setOnError(error);
+        throw error; // throw error to CatalogFunctionMixin
+      }
+    }
+
+    get refreshInterval() {
+      return 2;
+    }
+
+    private pollingForResults = false;
+
+    /**
+     * Called every refreshInterval
+     *
+     * @return true if job has finished, false otherwise
+     */
+    async pollForResults(): Promise<boolean> {
+      throw "pollForResults not implemented";
+    }
+
+    /**
+     * This function adapts AutoRefreshMixin's refreshData with this Mixin's pollForResults - adding the boolean return value which triggers refresh disable
+     */
+    refreshData() {
+      if (this.pollingForResults) {
+        return;
+      }
+
+      this.pollingForResults = true;
+
+      this.pollForResults()
+        .then(finished => {
+          if (finished) {
+            runInAction(() => {
+              this.setTrait(CommonStrata.user, "jobStatus", "finished");
+              this.setTrait(CommonStrata.user, "refreshEnabled", false);
+            });
+            this.onJobFinish(true);
+          }
+          this.pollingForResults = false;
+        })
+        .catch(error => {
+          runInAction(() => {
+            this.setTrait(CommonStrata.user, "jobStatus", "error");
+            this.setTrait(CommonStrata.user, "refreshEnabled", false);
+            this.setOnError(error);
+          });
+          this.pollingForResults = false;
+        });
+    }
+
+    private downloadingResults = false;
+
+    /**
+     * This handles downloading job results, it can be triggered three ways:
+     * - `_invoke` returns true {@link CatalogFunctionJobMixin#invoke}
+     * - `pollForResults` returns true {@link CatalogFunctionJobMixin#refreshData}
+     * - on `loadMetadata` if `jobStatus` is "finished", and `!downloadedResults`  {@link CatalogFunctionJobMixin#forceLoadMetadata}
+     */
+    private async onJobFinish(addToWorkbench = this.inWorkbench) {
+      // Download results when finished
+      if (
+        this.jobStatus === "finished" &&
+        !this.downloadedResults &&
+        !this.downloadingResults
+      ) {
+        this.downloadingResults = true;
+        this.results = (await this.downloadResults()) || [];
+        this.results.forEach(result => {
+          if (Mappable.is(result))
+            result.setTrait(CommonStrata.user, "show", true);
+          if (addToWorkbench) this.terria.workbench.add(result);
```

Comment:
I use this instead of `addToWorkbench()` because this is a `GroupMixin` and is therefore ignored in `addToWorkbench`

Reply:
I ended up reverting this to using `addToWorkbench`